File: 프로그래머스/lv1/12921. 소수 찾기/소수 찾기.js

let solution = (n) => {
    let arr = Array(n + 1).fill(true).fill(false, 0, 2);
 
    for (let i = 2; i < Number(n ** 0.5) + 1; i++) {
        if (arr[i]) {
            for (let j = i * i; j <= n; j += i) {
                arr[j] = false;
                
            }
        }
    }
let count = 0
    arr.map(item =>{
        if(item === true){
            count++
        }
    });
    return count
}

console.log(solution(10000));


// function solution(n) {
    
//     let count = 1; 
    
//     for(let i=2; i<=n; i++){ 
//          if(isPrime(i)) count++;
//     }


//     function isPrime(num){
//         if(num % 2 === 0) return false;
//         for(let i = 2; i < Math.sqrt(num)+1; i++){
//             if(num%i==0) return false; 
//         }
//         return true;
//     }

//     return count;
// }

File: 프로그래머스/lv1/12921. 소수 찾기/소수 찾기.cpp
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <math.h>

int solution(int n) {
    bool *prime_check = (bool *)calloc(n + 1, sizeof(bool));
    int answer = 0;

    for (int i = 2; i <= n; i++) {
        prime_check[i] = true;
    }

    for (int i = 2; i <= (int)sqrt(n); i++) {
        if (prime_check[i]) {
            for (int j = i * i; j <= n; j += i) {
                prime_check[j] = false;
            }
        }
    }

    for (int i = 2; i <= n; i++) {
        if (prime_check[i]) {
            answer++;
        }
    }

    free(prime_check);

    return answer;
}

int main() {
    int n = 7;
    int result = solution(n);
    printf("소수의 개수: %d\n", result);

    return 0;
}
